collections/_article/d3d12-shaders-with-a-twist.md

```diff
@@ -0,0 +1,112 @@
+
+> DXIL register = SPIR-V set * 100000000 + binding * 100000
+
+A couple of notes on this:
+- We can make such a liberal use of DXIL registers since they are logical, just a way of tagging the bindings. (In past versions that was not the case, as far as I can tell.)
+- Regarding the last multiplication, if we understand Vulkan numbered bindings as "slots" and so we do with Direct3D register indices, in principle there would be no reason for it. However, Vulkan slots can hold entire arrays, whereas D3D ones need one slot per array element. Therefore, with that multiplication we are reserving a big range of D3D bindings to account for however big a incoming Vulkan array may be.
+
+About mapping the Vulkan resource types (image, buffer, etc.) to their D3D12 counterparts (SRV, UAV, etc.), there's no 1:1 mapping. It depends on extra attributes that decorate each binding, and even on actual usage! With the SPIRV-Cross approach, the Godot D3D12 renderer needed to do reflection on the DXIL shader to find out (reflection was used for the bindings, too.) In contrast, with the NIR way, the callbacks mentioned earlier can also get information about which kind of Direct3D resource each binding is being translated to, on the go.
+
+In conclusion, the NIR way was kind of uncharted territory, but turned out to be much more convenient. I had to make a few PRs to Mesa to add or fix little pieces of translation logic to this part of Dozen, though. But again, that's a great learning experience.
+
+## Specialization constants (creatively working around the lack thereof in Direct3D)
+
+![SCs in DXIL](/storage/blog/d3d12/scs-in-dxil.jpg)
+
+For the first version of the PR (the SPIRV-Cross + DirectX Shader Compiler way) I published [a Twitter thread](https://twitter.com/RandomPedroJ/status/1532725156623286272?s=20) explaining how I approached this problem. Some parts are still true in the NIR-based PR. In any case, in this writeup I'm doing the full updated explanation of how it works now, so there's no need to read the other one.)
+
+Vulkan features specialization constants (SCs from now on), which are some sort of compile-time constants in terms of optimization of the final shader code that runs on the GPU (branch pruning, informed loop unrolling, etc.), yet much more flexible, in that they can be applied at runtime to create the graphics/compute pipelines with **shaders already compiled to SPIR-V**.
+
+In Direct3D the closest is actual constants or preprocessor definitions in the HLSL code. That means that, in order to create multiple pipelines corresponding to different variants of the same shader, you have to compile multiple versions of the HLSL code with different values set each time at the source code level.
```

Comment:
```suggestion
In Direct3D, the closest is actual constants or preprocessor definitions in the HLSL code. That means that, in order to create multiple pipelines corresponding to different variants of the same shader, you have to compile multiple versions of the HLSL code with different values set each time at the source code level.
```

Reply:
I've changed it to, "The closest in Direct3D is [...]." Good?